tests/regression/fn/fn_nested_reduce.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <gtest/gtest.h>
+
+#include <gridtools/fn.hpp>
+#include <gridtools/fn/backend/naive.hpp>
+
+#include "simple_mesh.hpp"
+
+using namespace gridtools;
+using namespace fn;
+using namespace literals;
+
+constexpr auto zero = []<class T>(T, auto...) { return T{}; };
+constexpr auto sum = reduce<plus, zero>;
+
+constexpr auto dot_fun = [](auto acc, auto x, auto y) { return acc + x * y; };
+constexpr auto dot = reduce<dot_fun, zero>;
+
+template <auto V2E>
+constexpr auto nb_sum = [](auto const &x) { return sum(shift<V2E>(x)); };
+
+template <auto V2V, auto V2E, bool UseTmp>
+constexpr auto nb_nb_sum = [](auto const &x) { return sum(shift<V2V>(lift<sum, UseTmp>(shift<V2E>(x)))); };
```

Comment:
to discuss

sum(sum(shift<V2V>(shift<V2E>(x)))) -> sum(sum(shift<V2E,V2V>(x)))

shift(O1)(lift(foo)(it)) == foo(shift(O1)(it))